docs: ADR-010 enabling standalone ICS-02 integration (#1116)
cosmos/ibc-rs: docs/architecture/adr-010-enable-standalone-ics02-integration.md

      client APIs, and client-specific APIs can get cleared. This ADR aims for a
      more transparent relationship between these contexts, making it easier to
      understand their interactions and roles. For instance some client relevant
      methods are located under the client contexts, while others are under the
      main (ICS-24) contexts.

3. **Minimizing associated types’ declarations**
    - Our current APIs pose a challenge by requiring users to introduce multiple
      yet the same associated types like AnyClientState, AnyConsensusState, etc
      across different layers. A more efficient solution involves introducing
      such types in fewer places, providing better access to the traits across
      different layers.

## Decision

Here is a high-level diagram that illustrates the optimal boundaries for each
context trait, specifying the scope of access by each of these contexts. In
addition, the classification of APIs should be such that they allow the
independent integration of light client implementations.

<div style="text-align: center;">
  <img src="./assets/adr010.png" alt="Main components" width="800" height="600">
</div>

> The primary `ValidationContext` and `ExecutionContext` traits will be

Reviewer: ```suggestion
The primary `ValidationContext` and `ExecutionContext` traits at the ICS-24 host level will be
```

restructured as follows, only having sufficient access to respective client
contexts, without caring about the client-specific types or methods. It is
noteworthy that, to better illustrate the desired outcome from the current
state, the code snippets below are in the `diff` format.

```diff
pub trait ValidationContext {
    type V: ClientValidationContext;
-    type E: ClientExecutionContext;
-    type AnyConsensusState: ConsensusState;
-    type AnyClientState: ClientState<Self::V, Self::E>;

    /// Retrieve the context that implements all clients' `ValidationContext`.
    fn get_client_validation_context(&self) -> &Self::V;

>      /// This method will be removed and instead, the `ClientStateDecoder` trait will be introduced.

Reviewer: ```suggestion
     // This method will be removed and replaced by a `ClientStateDecoder` trait that will encapsulate the ability to decode a client state from an `Any`
```

{
-    type V: ClientValidationContext;
-    type AnyClientState: ClientState<Self::V, Self>;
-    type AnyConsensusState: ConsensusState;
+    type ClientStateMut: ClientStateExecution<Self>;

+    fn client_state_mut(&self, client_id: &ClientId) -> Result<Self::ClientStateMut, ContextError> {
+        self.client_state(client_id)
+    }

    fn store_client_state(
        &mut self,
        client_state_path: ClientStatePath,
        client_state: Self::ClientStateMut,
    ) -> Result<(), ContextError>;

    fn store_consensus_state(
        &mut self,
        consensus_state_path: ClientConsensusStatePath,
        consensus_state: Self::ConsensusStateRef,
    ) -> Result<(), ContextError>;

    fn delete_consensus_state(
        &mut self,
        consensus_state_path: ClientConsensusStatePath,
    ) -> Result<(), ContextError>;

    fn store_update_meta(
        &mut self,
        client_id: ClientId,
        height: Height,
        host_timestamp: Timestamp,
        host_height: Height,
    ) -> Result<(), ContextError>;

    fn delete_update_meta(
        &mut self,
        client_id: ClientId,
        height: Height,
    ) -> Result<(), ContextError>;
}
```

> The introduction of the `ClientStateMut` associated type additional to the

Reviewer: ```suggestion
The introduction of the `ClientStateMut` associated type in addition to the
```

`ClientStateRef` became necessary to tackle the limitation that the
`ClientState` retrieved from the regular `client_state()` method provides access
only to validation methods. However, in `execute` handlers, there are scenarios
(like
[here](https://github.com/cosmos/ibc-rs/blob/f272f30e0f773d85a99fc553b75d41f9f768d5c5/ibc-core/ics02-client/src/handler/update_client.rs#L54))
where access to the execution methods of the client state is required. We aim to
simplify the user experience by providing a default implementation, relieving
users from the need to implement the `client_state_mut` method.

Also, the introduction of `<ClientStateRef = Self::ClientStateMut>` is prompted
by the need to address the characteristics of concrete `ClientState`
definitions. For instance, in the case of ICS-07, such as `TmClientState`, the
struct definition can't be split into two fragments, one for validation and the
other for execution. Therefore, contexts implementing `ClientExecutionContext`
must introduce a `ClientStateMut` type the same as `ClientStateRef`.

With the mentioned classification, we can now streamline ICS-07 specific APIs,

> eliminating the requirement for implementing a redundant consensus_state()

Reviewer: ```suggestion
eliminating the requirement for implementing a redundant `consensus_state()`
```

> method. For the sake of simplification, we can remove the CommonContext trait

Reviewer: ```suggestion
method. For the sake of simplification, we can remove the `CommonContext` trait
```

> and consolidate everything under the TmValidationContext as follows:

Reviewer: ```suggestion
and consolidate everything under the `TmValidationContext` as follows:
```

+impl<T> ExecutionContext for T where T: ValidationContext + ClientExecutionContext {}

```

### Remarks

- We move away from the `decode_client_state()` method. Since per our design,
  users must utilize a `ClientState` type that implements both `TryFrom<Any>`
  and `Into<Any>`, therefore, we can offer a trait named `ClientStateDecoder`
  encompassing a `decode_from_any()` method as follows, making it readily
  available for users seeking to decode a client state from `Any` type:

    ```rust
    pub trait ClientStateDecoder: TryFrom<Any, Error = ClientError> {
        fn decode_from_any(client_state: Any) -> Result<Self, ClientError> {
            Self::try_from(client_state)
        }
    }

    impl<T> ClientStateDecoder for T where T: TryFrom<Any, Error = ClientError> {}

    ```

- We will maintain the `client_counter()` and `increase_client_counter()` methods
  within the main context traits. This stems from the fact that light clients do
  not rely on the relative positions in their processes. Additionally, these
  counters are globally tracked, and only IBC handlers invoke these methods for
  setting or retrieving client identifiers.

> - The `host_consensus_state()` renamed to the `self_consensus_state()` for

Reviewer: ```suggestion
- The `host_consensus_state()` will be renamed to `self_consensus_state()` in order to maintain
```

      under the `ValidationContext`. This will be rectified to ensure that only
      the `ClientState` validation method is callable within the validation
      contexts. This enhances the precision of access control, aligning it more
      closely with the intended functionality.

2. **Bringing more clarity to how context APIs are related**
    - The ambiguous relationship between the main context APIs at ICS-24, ICS-02
      client APIs, and client-specific APIs can get cleared. This ADR aims for a
      more transparent relationship between these contexts, making it easier to
      understand their interactions and roles. For instance some client relevant
      methods are located under the client contexts, while others are under the
      main (ICS-24) contexts.

3. **Minimizing associated types’ declarations**
    - Our current APIs pose a challenge by requiring users to introduce multiple
      yet the same associated types like AnyClientState, AnyConsensusState, etc
      across different layers. A more efficient solution involves introducing
      such types in fewer places, providing better access to the traits across

>       different layers.

Reviewer 2: Can you please comment here on what features we need from rustc to improve this later?

- associated_type_bounds
- associated_type_defaults
- implied_bounds

Author: [034604a](https://github.com/cosmos/ibc-rs/pull/1116/commits/034604aca6663a554ea1704b01fbbd6dd899789f)
